llvm/lib/Analysis/ScalarEvolution.cpp

```diff
@@ -5933,18 +5925,21 @@ const SCEV *ScalarEvolution::createAddRecFromPHI(PHINode *PN) {
     // We can generalize this saying that i is the shifted value of BEValue
     // by one iteration:
     //   PHI(f(0), f({1,+,1})) --> f({0,+,1})
-    const SCEV *Shifted = SCEVShiftRewriter::rewrite(BEValue, L, *this);
-    const SCEV *Start = SCEVInitRewriter::rewrite(Shifted, L, *this, false);
-    if (Shifted != getCouldNotCompute() &&
-        Start != getCouldNotCompute()) {
-      const SCEV *StartVal = getSCEV(StartValueV);
-      if (Start == StartVal) {
-        // Okay, for the entire analysis of this edge we assumed the PHI
-        // to be symbolic.  We now need to go back and purge all of the
-        // entries for the scalars that use the symbolic expression.
-        forgetMemoizedResults(SymbolicName);
-        insertValueToMap(PN, Shifted);
-        return Shifted;
+
+    // Do not allow refinement in rewriting of BEValue.
+    if (isGuaranteedNotToCauseUB(BEValue)) {
```

Comment:
Not sure whether we need `isGuaranteedNotToBePoison` here. IIRC SCEV doesn't leverage poison to refine expressions:

> // The only way poison may be introduced in a SCEV expression is from a
// poison SCEVUnknown (ConstantExprs are also represented as SCEVUnknown,
// not SCEVConstant). Notably, nowrap flags in SCEV nodes can *not*
// introduce poison -- they encode guaranteed, non-speculated knowledge.

Reply:
Poison is still a problem if we, for example, replace `0` with `0 *  %a`. Maybe a bit tricky to get things in the right form so it doesn't simplify, but something like `phi(0, zext({1,+,1}) * %a)` (where the addrec is not nuw so zext doesn't fold) might do? The shifted form would be `zext({0,+,1}) * %a` and the shifted start would be `zext(0) * a` which is `0`. But we can't replace 0 by `zext({0,+,1}) * %a`, as it may be more poisonous.

Reply:
Done.